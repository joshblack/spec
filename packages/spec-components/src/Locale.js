/* @flow */

import React from 'react';
import PropTypes from 'prop-types';
import { IntlProvider, addLocaleData } from 'react-intl';

import type { Node } from 'react';

type Props = {
  language: string,
  locale?: string,
  loadMessages: (
    language: string,
    locale?: string
  ) => Promise<{ [key: string]: string }>,
  render: (error: ?Error, isLoading: boolean) => Node,
};

type State = {
  error: ?Error,
  messages: ?{ [key: string]: string },
  isLoading: boolean,
};

export default class Locale extends React.Component<Props, State> {
  static propTypes = {
    language: PropTypes.string.isRequired,
    locale: PropTypes.string,
    loadMessages: PropTypes.func.isRequired,
    render: PropTypes.func.isRequired,
  };

  state = {
    error: null,
    messages: null,
    isLoading: true,
  };

  componentDidMount() {
    const { language, locale, loadMessages } = this.props;
    const handleSuccess = ([messages]) => {
      this.setState({ error: null, isLoading: false, messages });
    };
    const handleError = error => {
      this.setState({ error, isLoading: false });
    };

    Promise.all([loadMessages(language, locale), this.loadLocaleData(language)])
      .then(handleSuccess, handleError)
      .catch(handleError);
  }

  loadLocaleData = (language: string) =>
    // $FlowFixMe
    import(`react-intl/locale-data/${language}`).then(localeData => {
      addLocaleData(localeData);
    });

  render() {
    const { language, locale, render } = this.props;
    const { error, isLoading, messages } = this.state;

    if (error !== null || isLoading) {
      return render(error, isLoading);
    }

    const providerLocale = locale === undefined ? language : locale;

    return (
      <IntlProvider locale={providerLocale} messages={messages}>
        {render(error, isLoading)}
      </IntlProvider>
    );
  }
}
