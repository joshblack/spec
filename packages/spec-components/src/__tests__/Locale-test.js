/* @noflow */

import React from 'react';
import renderer from 'react-test-renderer';
import Locale from '../Locale';

describe('Locale component', () => {
  it('should render', done => {
    const tree = renderer.create(
      <Locale
        language="en"
        locale="en-US"
        loadMessages={() =>
          Promise.resolve({
            greeting: 'hello',
          })}
        render={(error, isLoading) => {
          if (isLoading) {
            return <p>Loading...</p>;
          }

          if (error) {
            return <p>Error</p>;
          }

          return <h1>Hi there</h1>;
        }}
      />
    );

    setTimeout(() => {
      expect(tree.toJSON()).toMatchSnapshot();
      done();
    });
  });
});
