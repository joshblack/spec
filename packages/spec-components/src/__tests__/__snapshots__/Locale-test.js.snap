// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Locale component should render 1`] = `
<h1>
  Hi there
</h1>
`;
